design/restic-backup-and-restore-progress.md

```diff
@@ -0,0 +1,114 @@
+# Progress reporting for restic backups and restores
+
+Status: Draft
+
+During long-running restic backups/restores, there is no visibility into what (if anything) is happening, making it hard to know if the backup/restore is making progress or hung, how long the operation might take, etc.
+We should capture progress during restic operations and make it user-visible so that it's easier to reason about.
+This document proposes an approach for capturing progress of backup and restore operations and exposing this information to users.
+
+## Goals
+
+- Provide basic visibility into restic operations to inform users about their progress.
+
+## Non Goals
+
+- Capturing progress for non-restic backups and restores.
+
+## Background
+
+(Omitted, see introduction)
+
+## High-Level Design
+
+### restic backup progress
+
+The `restic backup` command provides progress reporting to stdout in JSON format, which includes the completion percentage of the backup.
+This progress will be read on some interval and the PodVolumeBackup Custom Resource's (CR) status will be updated with this information.
+
+### restic restore progress
+
+The `restic ls` command returns the number of files in a backup, including the size of each file.
+This can be compared with the total size of these files in the volume periodically to calculate the completion percentage of the restore.
+The PodVolumeRestore CR's status will be updated with this information.
+
+## Detailed Design
+
+## Changes to PodVolumeBackup and PodVolumeRestore Status type
+
+A new `Progress` field will be added to PodVolumeBackupStatus and PodVolumeRestoreStatus of type `PodVolumeOperationProgress`:
+
+```
+type PodVolumeOperationProgress struct {
+  TotalBytes int64
+  BytesDone int64
+}
+```
+
+### restic backup progress
+
+restic added support for [streaming JSON output for the `restic backup` command](https://github.com/restic/restic/pull/1944) in 0.9.5.
+Our current images ship restic 0.9.4, and so the Dockerfile will be updated to pull the new version: https://github.com/heptio/velero/blob/af4b9373fc73047f843cd4bc3648603d780c8b74/Dockerfile-velero#L21.
+With the `--json` flag, `restic backup` outputs single lines of JSON reporting the status of the backup:
+
+```
+{"message_type":"status","percent_done":0,"total_files":1,"total_bytes":21424504832}
```

Comment:
I notice this includes a `percent_done` field. Is this excluded in order to provide symmetry for backup and restore?

Reply:
I was originally going to use this, but we also want to report the TotalBytes and BytesDone in status for both restore and backup so we can ignore `percent_done` and just calculate that in the Velero CLI.